compiler/rustc_macros/src/diagnostics/subdiagnostic_builder.rs

```diff
@@ -0,0 +1,612 @@
+    /// `#[multipart_suggestion{,_short,_hidden,_verbose}]`
+    MultiPartSuggestion { suggestion_kind: SubdiagnosticSuggestionKind },
+}
+
+impl quote::IdentFragment for SubdiagnosticKind {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            SubdiagnosticKind::Label => write!(f, "label"),
+            SubdiagnosticKind::Note => write!(f, "note"),
+            SubdiagnosticKind::Help => write!(f, "help"),
+            SubdiagnosticKind::Warn => write!(f, "warn"),
+            SubdiagnosticKind::Suggestion { .. } => write!(f, "suggestion_with_style"),
+            SubdiagnosticKind::MultiPartSuggestion { .. } => {
+                write!(f, "multipart_suggestion_with_style")
+            }
+        }
+    }
+
+    fn span(&self) -> Option<proc_macro2::Span> {
+        None
+    }
+}
+
+/// Tracks persistent information required for building up the call to add to the diagnostic
+/// for the final generated method. This is a separate struct to `SessionSubdiagnosticDerive`
+/// only to be able to destructure and split `self.builder` and the `self.structure` up to avoid a
+/// double mut borrow later on.
+pub(super) struct SessionSubdiagnosticDeriveBuilder<'a> {
+    /// The identifier to use for the generated `DiagnosticBuilder` instance.
+    pub diag: &'a syn::Ident,
+
+    /// Info for the current variant (or the type if not an enum).
+    pub variant: &'a VariantInfo<'a>,
+    /// Span for the entire type.
+    pub span: proc_macro::Span,
+
+    /// Store a map of field name to its corresponding field. This is built on construction of the
+    /// derive builder.
+    pub fields: HashMap<String, TokenStream>,
+
+    /// Identifier for the binding to the `#[primary_span]` field.
+    pub primary_span_field: Option<(proc_macro2::Ident, proc_macro::Span)>,
+
+    /// The applicability of a `#[suggestion]` or `#[multipart_suggestion]`
+    pub applicability: Option<(TokenStream, proc_macro::Span)>,
+
+    ///// Identifiers and replacement string for the binding to the `#[suggestion_part]` fields.
+    //pub suggestions: Vec<(proc_macro2::Ident, TokenStream)>,
+    pub has_suggestion_parts: bool,
+}
+
+impl<'a> HasFieldMap for SessionSubdiagnosticDeriveBuilder<'a> {
+    fn get_field_binding(&self, field: &String) -> Option<&TokenStream> {
+        self.fields.get(field)
+    }
+}
+
+impl<'a> SessionSubdiagnosticDeriveBuilder<'a> {
+    fn identify_kind(
+        &mut self,
+    ) -> Result<Option<(SubdiagnosticKind, Path)>, DiagnosticDeriveError> {
+        let mut kind_slug = None;
+
+        for attr in self.variant.ast().attrs {
+            let span = attr.span().unwrap();
+
+            let name = attr.path.segments.last().unwrap().ident.to_string();
+            let name = name.as_str();
+
+            let meta = attr.parse_meta()?;
+            let Meta::List(MetaList { ref nested, .. }) = meta else {
+                throw_invalid_attr!(attr, &meta);
+            };
+
+            enum KindTag {
```

Comment:
```suggestion
            enum SubdiagnosticKind {
```
nit: I prefer this name :)

Reply:
This was a holdover from when I stored more information in the "real" `SubdiagnosticKind`, so I needed a separate "tag" enum that encodes just the type of diagnostic - I've rewritten it to make that unnecessary.